laterpay/application/Controller/Frontend/Post.php

```diff
@@ -496,7 +496,105 @@ public function check_user_access( $has_access, $post_id = null ) {
     }
 
     /**
-     * Prefetch the post access for posts in the loop.
+     * Encrypt image source to prevent direct access.
+     *
+     * @wp-hook wp_get_attachment_image_attributes
+     *
+     * @param array        $attr Attributes for the image markup
+     * @param WP_Post      $post Image attachment post
+     * @param string|array $size Requested size
+     *
+     * @return mixed
+     */
+    public function encrypt_image_source( $attr, $post, $size ) {
+        $is_purchasable = LaterPay_Helper_Pricing::is_purchasable( $post->ID );
+        if ( $is_purchasable ) {
+            $access         = LaterPay_Helper_Post::has_access_to_post( $post );
+            $attr['src']    = LaterPay_Helper_File::get_encrypted_resource_url(
+                $post->ID,
+                $attr['src'],
+                $access,
+                'attachment'
+            );
+        }
+
+        return $attr;
+    }
+
+    /**
+     * Encrypt attachment URL to prevent direct access.
+     *
+     * @wp-hook wp_get_attachment_url
+     *
+     * @param string $url     URL for the given attachment
+     * @param int    $post_id Attachment ID
+     *
+     * @return string
+     */
+    public function encrypt_attachment_url( $url, $post_id ) {
+        // get current post
+        $current_post = get_post();
+        if ( $current_post === null ) {
+            return $url;
+        }
+
+        $is_purchasable = LaterPay_Helper_Pricing::is_purchasable( $current_post->ID );
+        if ( $is_purchasable && $current_post->ID === $post_id ) {
+            $access = LaterPay_Helper_Post::has_access_to_post( $current_post );
+
+            // prevent from exec, if attachment is an image and user does not have access
+            if ( ! $access && strpos( $current_post->post_mime_type, 'image' ) !== false ) {
+                return '';
+            }
+
+            // encrypt attachment URL
+            $url = LaterPay_Helper_File::get_encrypted_resource_url(
+                $post_id,
+                $url,
+                $access,
+                'attachment'
+            );
+
+            return $url;
+        }
+
+        return $url;
+    }
+
+    /**
+     * Prevent attachments from prepend paid content.
```

Comment:
By default wordpress render attachment right before content, so it not included in the content and not covered by laterpay overlay/teaser. User can access this attachment easily by clicking on it. We should prevent such behaviour if this attachment is paid and user has no access to it. Filter "prepend_attachment" do this.